Issue #91 - Ability to specify canaryService Service object to reach only canary pods

alexmt opened the pull request:
Closes #91

dthomson25 commented on controller/service.go:
```diff
@@ -155,6 +156,35 @@ func (c *Controller) getPreviewAndActiveServices(r *v1alpha1.Rollout) (*corev1.S
+			svc.Spec.Selector = make(map[string]string)
+		}
+
+		hash := controller.ComputeHash(&r.Spec.Template, r.Status.CollisionCount)
```
Instead of calculating the hash from the template, we should use calculate the hash from the newRS podSpec to protect against the k8s dependency change (#88).

alexmt replied:
Done.

dthomson25 commented on controller/service.go:
```diff
@@ -155,6 +156,35 @@ func (c *Controller) getPreviewAndActiveServices(r *v1alpha1.Rollout) (*corev1.S
+
+		hash := controller.ComputeHash(&r.Spec.Template, r.Status.CollisionCount)
+		if svc.Spec.Selector[v1alpha1.DefaultRolloutUniqueLabelKey] != hash {
+			patch := fmt.Sprintf(switchSelectorPatch, v1alpha1.DefaultRolloutUniqueLabelKey, hash)
```
Can you use the c.switchServiceSelector method to patch the service?

alexmt replied:
done

dthomson25 commented on controller/canary_test.go:
```diff
@@ -896,3 +898,52 @@ func TestCanaryRolloutStatusHPAStatusFields(t *testing.T) {
+func TestCanaryRolloutWithCanaryService(t *testing.T) {
+	f := newFixture(t)
+
+	canarySvc := newService("canary", 80, make(map[string]string))
```
What is there a reason that you instantiate the selector to nothing? If we don't, that will allow us to also test the case where the selector is nil.

alexmt replied:
Good point. Changed `make(map[string]string)` to nil.

dthomson25 commented on controller/service.go:
```diff
@@ -155,6 +156,35 @@ func (c *Controller) getPreviewAndActiveServices(r *v1alpha1.Rollout) (*corev1.S
+				newStatus := r.Status.DeepCopy()
+				cond := conditions.NewRolloutCondition(v1alpha1.RolloutProgressing, corev1.ConditionFalse, conditions.ServiceNotFoundReason, msg)
+				conditions.SetRolloutCondition(newStatus, *cond)
+				c.persistRolloutStatus(r, newStatus, &r.Spec.Paused)
```
I'm wondering if we should abstract this logic into a separate method because we have this code repeated 3 times. Thoughts?

alexmt replied:
good catch! done

dthomson25 commented on controller/service.go:
```diff
@@ -234,6 +264,10 @@ func getRolloutServiceKeys(rollout *v1alpha1.Rollout) []string {
 		if rollout.Spec.Strategy.BlueGreenStrategy.PreviewService != "" {
 			servicesSet[fmt.Sprintf("%s/%s", rollout.Namespace, rollout.Spec.Strategy.BlueGreenStrategy.PreviewService)] = true
 		}
+	} else if rollout.Spec.Strategy.CanaryStrategy != nil {
```
Can we add another test to TestGetRolloutServiceKeysForCanary to cover when a canaryService specified?

alexmt replied:
done

jessesuen commented on controller/service.go:
```diff
@@ -155,6 +156,35 @@ func (c *Controller) getPreviewAndActiveServices(r *v1alpha1.Rollout) (*corev1.S
 }
 
+func (c *Controller) reconcileCanaryService(r *v1alpha1.Rollout, newRS *appsv1.ReplicaSet) error {
+	if r.Spec.Strategy.CanaryStrategy != nil && r.Spec.Strategy.CanaryStrategy.CanaryService != "" {
```
Style nit, we would save a block of indentation with:
```golang
if r.Spec.Strategy.CanaryStrategy == nil || r.Spec.Strategy.CanaryStrategy.CanaryService == "" {
    return nil
}
```

alexmt replied:
done

alexmt commented:
Addressed review comments. Also discovered that tests became flaky after merging https://github.com/argoproj/argo-rollouts/pull/95: additional service index requires run syncronize informers in test. The https://github.com/argoproj/argo-rollouts/pull/100/commits/9919549b147684e787414ddfb68c2e4f73f79941 fixes it

dthomson25 commented on controller/service.go:
```diff
@@ -3,17 +3,16 @@ package controller
 
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/conditions"
+	logutil "github.com/argoproj/argo-rollouts/utils/log"
```
Style nix: can we move this back below the external lib imports

alexmt replied:
done
